packages/flutter/test/foundation/service_extensions_test.dart

```diff
@@ -0,0 +1,296 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:typed_data';
+import 'dart:ui' as ui;
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/gestures.dart';
+import 'package:flutter/rendering.dart';
+import 'package:flutter/scheduler.dart';
+import 'package:flutter/services.dart';
+import 'package:flutter/widgets.dart';
+import 'package:test/test.dart';
+
+class TestServiceExtensionsBinding extends BindingBase
+  with SchedulerBinding,
+       ServicesBinding,
+       GestureBinding,
+       RendererBinding,
+       WidgetsBinding {
+
+  final Map<String, ServiceExtensionCallback> extensions = <String, ServiceExtensionCallback>{};
+
+  @override
+  void registerServiceExtension({
+    @required String name,
+    @required ServiceExtensionCallback callback
+  }) {
+    expect(extensions.containsKey(name), isFalse);
+    extensions[name] = callback;
+  }
+
+  Future<Map<String, String>> testExtension(String name, Map<String, String> arguments) {
+    expect(extensions.containsKey(name), isTrue);
+    return extensions[name](arguments);
+  }
+
+  int reassembled = 0;
+  @override
+  Future<Null> reassembleApplication() {
+    reassembled += 1;
+    return super.reassembleApplication();
+  }
+
+  bool frameScheduled = false;
+  @override
+  void scheduleFrame() {
+    frameScheduled = true;
+  }
+  void doFrame() {
+    frameScheduled = false;
+    if (ui.window.onBeginFrame != null)
+      ui.window.onBeginFrame(const Duration());
+  }
+
+  Future<Null> flushMicrotasks() {
+    Completer<Null> completer = new Completer<Null>();
+    new Timer(const Duration(), () {
+      completer.complete();
+    });
+    return completer.future;
+  }
+}
+
+void main() {
+  test('Service Extension Tests', () async {
+    TestServiceExtensionsBinding binding = new TestServiceExtensionsBinding();
+
+    expect(binding.frameScheduled, isTrue);
+    binding.doFrame(); // initial frame scheduled by creating the binding
+    expect(binding.frameScheduled, isFalse);
+
+    Map<String, String> result;
+    Future<Map<String, String>> pendingResult;
+    bool completed;
+
+    expect(debugPrint, equals(debugPrintThrottled));
+    List<String> console = <String>[];
+    debugPrint = (String message, { int wrapWidth }) {
+      console.add(message);
+    };
+
+    // the following list is alphabetical, one paragraph per extension
```

Comment:
has to be in a test because it has expectations as well.

but i split it into 13 tests, a pretest, 11 tests, a posttest. i hope the order is guaranteed!